### PR TITLE
Add NVTX special handling for Windows

### LINKED ISSUE #55: NVTX discovery on Windows
https://github.com/maleadt/NVTX.jl/pull/1/files#r210869873

### PR DESCRIPTION
**Pull request by musm**
resolve https://github.com/JuliaGPU/CUDAapi.jl/issues/55

### CODE REVIEW

**Comment by musm** [comment]
@maleadt  Thoughts on this?

**Comment by maleadt** [comment]
Not sure... The intent of `CUDA_PATH` env overrides is to point to a presumably non-systemwide CUDA installation, but then your NVTX special casing would still pick up the systemwide copy. Then again, I'm not sure it's even possible to have a non-systemwide CUDA installation on Windows. Having `find_toolkit` return NVTX-specific paths seems wrong, but is more of an issue with how NVIDIA doesn't install it under the a CUDA root.

**Comment by musm** [comment]
Right I'm  using the default according to:
https://docs.nvidia.com/gameworks/index.html#developertools/desktop/nsight/nvtx_library.htm 

So I went ahead and looked at the installers. The cuda installer doesn't even allow you to specify where you want NVTX to be installed to. It only  allows you to specify where you want the toolkit installed.  I'm assuming, regardless of where the toolkit is installed, NVTX is always installed to the directory specified in the documentation.  

So basically NVTX is installed in the same directory as the display drivers **which also cannot be modified**.  I.e. the location of the display drivers is fixed and NVTX is installed here as well. 

This should be a reasonable assumption.

**Comment by musm** [comment]
Is that at all clear?
In other words the Nvidia display drivers are installed to a fixed **unchangeable** location which is also where NVTX is installed to by the CUDA installer.

**Comment by musm** [comment]
I guess the only weird thing is that this is adding the location to a function called `cuda_toolkit`, even though it's more of a helper. But at the same time I'm guessing in linux systems it's installed in the same location, so maybe it's fine, with the understanding that `cuda_toolkit` finds also helper libraries.

**Comment by musm** [comment]
Actually, we should however first probe the NVTOOLSEXT environment variable, then check the default path.

**Comment by musm** [comment]
bors try

**Comment by maleadt** [comment]
bors try

**Comment by musm** [comment]
Added a test for NVTX, could we retrigger bors? Sorry about that.

**Comment by maleadt** [comment]
Sure, it's a shame I can't configure bors to allow everybody to at least try.
bors try

**Comment by musm** [comment]
Interesting  library fails to be found for >= cuda 9.2   on linux

**Comment by musm** [comment]
@maleadt  Do you have any thoughts on the linux test failures. Is NVTX  not included by default in the cuda installer on linux?

**Comment by maleadt** [comment]
> Is NVTX not included by default in the cuda installer on linux?

It is, and this branch works on my system. I'll have a look.

**Comment by maleadt** [comment]
bors try

**Comment by maleadt** [comment]
bors try

**Comment by maleadt** [comment]
bors try

**Comment by maleadt** [comment]
bors try

**Comment by musm** [comment]
bors try

**Comment by musm** [comment]
Is it not possible to just integrate CI on every push?

**Comment by maleadt** [comment]
No, gitlab does not support running ci on forks. Bors makes this possible, but its permission system is a little annoying.bors tryOn 30 Apr 2019 20:51, "Mustafa M." <notifications@github.com> wrote:Is it not possible to just integrate CI on every push?

**Comment by maleadt** [comment]
No, gitlab does not support running ci on forks. Bors makes this possible, but its permission system is a little annoying.
bors try

**Comment by musm** [comment]
Ok NP,  it's just annoying that I have to bother you test things out  😛

**Comment by musm** [comment]
CI looks good after updating  appveyor. The only issue is with   CUDA 8. The error is a bit mystifying:

```
Program 'cuda.exe' failed to run: The specified executable is not a valid 
application for this OS platform.At C:\projects\cudaapi-jl\.appveyor.ps1:27 
char:1
+ .\cuda.exe -s @components | Out-Null
+ ~~~~~~~~~~~~~~~~~~~~~~~~~.
At C:\projects\cudaapi-jl\.appveyor.ps1:27 char:1
+ .\cuda.exe -s @components | Out-Null
```

**Comment by maleadt** [comment]
What comes before is more relevant:

```
Start-FileDownloadInternal : Error downloading remote file: One or more errors 
occurred.
Inner Exception: The remote name could not be resolved: 'developer.nvidia.com'
```

Let's just try again
bors try

**Comment by musm** [comment]
Ah thanks, looks like it will now work.  BTW is there a reason  to  set   `Set-PSDebug -Trace 1`? The output is very verbose and makes it hard to read the output.

**Comment by musm** [comment]
Should we tag a new release after this PR?

**Comment by maleadt** [comment]
> BTW is there a reason to set `Set-PSDebug -Trace 1`? The output is very verbose and makes it hard to read the output.

It's the (less readable) equivalent of running with `bash -x`, and did help when developing the CI scripts. It should probably go now...



> Should we tag a new release after this PR?

I'm going to have a look at the macOS failure too, but I'll tag a release just after that.

Thanks for the contribution!

**Comment by musm** [comment]
thanks
